### PR TITLE
Initial configuration for appliance

### PR DESCRIPTION
**Pull request by jhernand**
This patch adds the initial configuration and jobs for the appliance project. These will build the binaries and run the linter.

### CODE REVIEW

**Comment by jhernand** [comment]
/pj-rehearse pull-ci-openshift-appliance-master-images pull-ci-openshift-appliance-master-lint

**Comment by jhernand** [comment]
/pj-rehearse pull-ci-openshift-appliance-master-images pull-ci-openshift-appliance-master-lint

**Comment by jhernand** [comment]
@nmagnezi @danielerez please review. Note that currently there are only two jobs specific for this project: [building the images](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/39812/rehearse-39812-pull-ci-openshift-appliance-master-images/1663533712567963648) and [running the linter](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/39812/rehearse-39812-pull-ci-openshift-appliance-master-lint/1663534635608444928). Both are passing.

**Comment by jhernand** [comment]
@bparees the bot says to ask you for approval, can you take a look?

**Review by danilo-gemoli — APPROVED** [review]

**Comment by bparees** [comment]
what actually goes into the "openshift-appliance" image?  what will it be used for?  And will there be more images coming that are associated with the appliance?

**Comment by jhernand** [comment]
> what actually goes into the "openshift-appliance" image? what will it be used for? And will there be more images coming that are associated with the appliance?

@bparees the "openshift-appliance" image is part of the effort to create a mechanism to install OpenShift as an appliance. The image will contain a command line tool that customers and partners will use to configure and create the appliance. The output of that tool will be a disk image containing the installer and all the required images, including custom images needed by the customer. That disk image will then be copied to the disks of the nodes of the cluster. When the nodes of the cluster are booted from those disks the installation will performed and completed automatically.

The development of the command line tool is happening here: https://github.com/openshift/appliance

Currently we don't expect to need more images for this project, but it is early in the development process, so that could change.

@oourfali @nmagnezi @danielerez feel free to correct me or provide more information.

**Comment by danielerez** [comment]
> > what actually goes into the "openshift-appliance" image? what will it be used for? And will there be more images coming that are associated with the appliance?
> 
> @bparees @dhellmann the "openshift-appliance" image is part of the effort to create a mechanism to install OpenShift as an appliance. The image will contain a command line tool that customers and partners will use to configure and create the appliance. The output of that tool will be a disk image containing the installer and all the required images, including custom images needed by the customer. That disk image will then be copied to the disks of the nodes of the cluster. When the nodes of the cluster are booted from those disks the installation will performed and completed automatically.
> 
> The development of the command line tool is happening here: https://github.com/openshift/appliance
> 
> Currently we don't expect to need more images for this project, but it is early in the development process, so that could change.

Just one note about the image, we would need to push it to quay.io later on, probably under 'edge-infrastructure' organization. I.e. quay.io/edge-infrastructure/openshit-appliance:latest

> 
> @oourfali @nmagnezi @danielerez feel free to correct me or provide more information.

**Comment by jhernand** [comment]
> Have you already worked out with the ART team that they're willing to manage building this image? It's not going into the release payload, right?

@dhellmann I am not aware of any conversation with the ART team. Can you point me to the right person or channel to start that conversation?

File: ci-operator/config/openshift/appliance/openshift-appliance-master.yaml
Comment (jhernand) [reply ci-operator/config/openshift/appliance/openshift-appliance-master.yaml]: Done.

**Comment by dhellmann** [comment]
> > Have you already worked out with the ART team that they're willing to manage building this image? It's not going into the release payload, right?
> 
> @dhellmann I am not aware of any conversation with the ART team. Can you point me to the right person or channel to start that conversation? My understanding is that this doesn't need to be part of the release payload.

@sosiouxme and/or #aos-art on slack

**Comment by jhernand** [comment]
@dhellmann @bparees after discussing this with @sosiouxme we agreed that this doesn't need to be manage building this image. We will still like to build and test it with prow. Is there something in this pull request incompatible with that?

**Comment by bparees** [comment]
> @dhellmann @bparees after discussing this with @sosiouxme we agreed that ART doesn't need to be manage building this image. We will still like to build and test it with prow. Is there something in this pull request incompatible with that?

this PR is compatible w/ that usage, but i would like to understand the long term intent of this image if it's not going to be built+published by ART?

There are basically two kinds of images defined/built here:

1) payload content (including ART-built OLM operators that aren't in the payload but do ship w/ OCP)
2) utility/test images used to run e2es or (rarer) to help support some sort of testing

this image doesn't quite seem to fit into either category?

**Comment by jhernand** [comment]
> this PR is compatible w/ that usage, but i would like to understand the long term intent of this image if it's not going to be built+published by ART?

My understanding from the conversation we had with the ART team (@sosiouxme feel free to correct me) is that it doesn't make much sense to have ART building and publishing images that aren't part of the OCP release or other operator released with OCP.

The long term purpose of this image is to contain the command like tool that will be used to to create the appliances. That will be somewhat similar to `openshift-install agent create image ...` but instead of creating an ISO image will create a raw disk image to be written to disk.

Not sure where we want to publish that longterm, but for the time being we intend to publish it to https://quay.io/organization/edge-infrastructure, like we do for the components of the assisted installer family. So that will be the upstream home, and we don't have finalized plans to have a downstream home.

> There are basically two kinds of images defined/built here:
> 
>     1. payload content (including ART-built OLM operators that aren't in the payload but do ship w/ OCP)
> 
>     2. utility/test images used to run e2es or (rarer) to help support some sort of testing
> 
> this image doesn't quite seem to fit into either category?

I'd say that this image doesn't strictly fit into any of those two categories, as it isn't tied to a particular OCP version and isn't an operator or a testing tool. It would be more similar to `oc-mirror`, for example. As I said our short term plan is to publish it to the same quay.io repository where we publish the components of the assited installer family.

**Comment by bparees** [comment]
ok..it all seems reasonable to me.  I think @dhellmann may still have some concerns over the use of "appliance" here, though, so i'll let him work with you on that if he's going to push for a change.

**Comment by dhellmann** [comment]
> ok..it all seems reasonable to me. I think @dhellmann may still have some concerns over the use of "appliance" here, though, so i'll let him work with you on that if he's going to push for a change.

Yeah, I'm going to be the bad guy, I'm afraid. The term "appliance" is _very_ overloaded, both in what people think it is and the number of different times we have had projects try to implement "an appliance". Let's see if we can come up with a more descriptive name that doesn't include the word "appliance" at all. Ideally we could rename the git repo to match the new name.

Some relevant terms based on the README seem to be "agent" and "image builder". Those are not really sufficiently distinct from what the agent installer does, though, so what other properties make this new thing unique?

**Comment by zaneb** [comment]
Something like `agent-preinstall-image-builder`? Something that is as purely descriptive of what it _does_ as possible, rather than tied to any particular use case, would be the ideal.

**Comment by oourfali** [comment]
@dhellmann we are in sync with product management on the appliance work, so that would be an official offering we provide. Also, as opposed to an "agent image builder", the resulting image will also contain the partner images.

@dhellmann Please reconsider it. 
Having said that, I don't want the naming to block this from getting in, so if you strongly think it isn't the proper name, we'll change it.

**Comment by bparees** [comment]
> How about just preinstall-image-builder as a compromise?

one of the goals is to name the images so they group logically together when sorted alphabetically.  Are we going to have other "preinstall" prefixed images, associated w/ this image?  Should this image be logically grouped/associated w/ any of our existing prefixes?

**Comment by dhellmann** [comment]
> > How about just preinstall-image-builder as a compromise?
> 
> one of the goals is to name the images so they group logically together when sorted alphabetically. Are we going to have other "preinstall" prefixed images, associated w/ this image? Should this image be logically grouped/associated w/ any of our existing prefixes?

Probably not. Does that mean bringing agent back into the name for consistency, or do you have another suggestion for the prefix?

**Comment by bparees** [comment]
> Does that mean bringing agent back into the name for consistency, or do you have another suggestion for the prefix?

i liked the agent prefix for this reason, but I didn't want to assume too much about the intent of the image.  The length doesn't seem terrible to me

**Comment by danielerez** [comment]
> bparees

Maybe `agent-install-image-builder` could work then?
I.e. it sounds a bit more descriptive with what the image does, which is installing a cluster. Rather than 'preinstall' which might be implying that the image is basically a preinstalled cluster(?)

**Comment by dhellmann** [comment]
> > bparees
> 
> Maybe `agent-install-image-builder` could work then? I.e. it sounds a bit more descriptive with what the image does, which is installing a cluster. Rather than 'preinstall' which might be implying that the image is basically a preinstalled cluster(?)

The agent installer works by building an image, so I think @zaneb was proposing "preinstall" instead of "install" to differentiate what this new tool builds.

**Comment by zaneb** [comment]
> The length doesn't seem terrible to me

Yeah, we already have `baremetal-hardware-event-proxy-operator-bundle` and - my personal favourite - `cluster-cloud-controller-manager-operator` (bingo), to say nothing of `cluster-resource-override-admission-operator-registry` :slightly_smiling_face:

**Comment by oourfali** [comment]
Let's settle for agent-preinstall-image-builder, as seems like there is a consensus on that.

**Comment by jhernand** [comment]
> Let's settle for agent-preinstall-image-builder, as seems like there is a consensus on that.

Updated the pull request to use `agent-preinstall-image-builder`, please review again.

**Comment by jhernand** [comment]
/pj-rehearse pull-ci-openshift-appliance-master-images pull-ci-openshift-appliance-master-lint

**Comment by oourfali** [comment]
@dhellmann do we have a green light?

**Comment by dhellmann** [comment]
> @dhellmann do we have a green light?

The name looks fine. The linter job is failing.

**Review by dhellmann — COMMENTED** [review]
Thanks for your patience. The name looks good. I'll leave it up to your team to review the other settings.

/approve

**Pull request merged**
